sdks/java/core/src/main/java/org/apache/beam/sdk/options/ValueProvider.java

```diff
@@ -0,0 +1,226 @@
+      T defaultValue, Long optionsId) {
+      this.methodName = methodName;
+      this.klass = klass;
+      this.defaultValue = defaultValue;
+      this.optionsId = optionsId;
+    }
+
+    static void setRuntimeOptions(PipelineOptions runtimeOptions) {
+      optionsMap.put(runtimeOptions.getOptionsId(), runtimeOptions);
+    }
+
+    @Override
+    public T get() {
+      PipelineOptions options = optionsMap.get(optionsId);
+      if (options == null) {
+        if (defaultValue != null) {
+          return defaultValue;
+        }
+        throw new RuntimeException("Not called from a runtime context.");
+      }
+      try {
+        Method method = klass.getMethod(methodName);
+        PipelineOptions methodOptions = options.as(klass);
+        InvocationHandler handler = Proxy.getInvocationHandler(methodOptions);
+        return ((ValueProvider<T>) handler.invoke(methodOptions, method, null)).get();
+      } catch (Throwable e) {
+        throw new RuntimeException("Unable to load runtime value.", e);
+      }
+    }
+
+    @Override
+    public boolean shouldValidate() {
+      return defaultValue != null;
+    }
+  }
+
+  /**
+   * Serializer for {@link ValueProvider}.
+   */
+  static class Serializer extends JsonSerializer<ValueProvider<?>> {
+    @Override
+    public void serialize(ValueProvider<?> value, JsonGenerator jgen,
+                          SerializerProvider provider) throws IOException {
+      if (value.shouldValidate()) {
+        jgen.writeObject(value.get());
+      } else {
+        jgen.writeNull();
+      }
+    }
+  }
+
+  /**
+   * Deserializer for {@link ValueProvider}, which handles type marshalling.
+   */
+  static class Deserializer extends JsonDeserializer<ValueProvider<?>>
+    implements ContextualDeserializer {
+
+    private final JavaType innerType;
+
+    // A 0-arg constructor is required.
+    Deserializer() {
+      this.innerType = null;
+    }
+
+    Deserializer(JavaType innerType) {
+      this.innerType = innerType;
+    }
+
+    @Override
+    public JsonDeserializer<?> createContextual(DeserializationContext ctxt,
+                                                BeanProperty property)
+        throws JsonMappingException {
+      checkNotNull(ctxt);
+      JavaType type = checkNotNull(ctxt.getContextualType());
+      JavaType[] params = type.findTypeParameters(ValueProvider.class);
+      if (params.length != 1) {
+        throw new RuntimeException(
+          "Unable to derive type for ValueProvider: " + type.toString());
+      }
+      JavaType param = params[0];
+      return new Deserializer(param);
+    }
+
+    @Override
+    public ValueProvider<?> deserialize(JsonParser jp, DeserializationContext ctxt)
+        throws IOException, JsonProcessingException {
+      JsonDeserializer dser = ctxt.findRootValueDeserializer(
+        checkNotNull(innerType));
+      Object o = dser.deserialize(jp, ctxt);
+      return StaticValueProvider.of(o);
```

Comment:
don't love all the `checkNotNull` calls with no error message. A runner may get this wrong, and we want to provide them helpful outs.

A sample improvement: I'm thinking that the check at L221 should actually be something like:

```
checkState(innerType != null, "Invalid %s: innerType is null. Serialization error?", getClass());
```

This gives a better error type (`IllegalStateException`, which conveys that the arguments to the function are not the issue) and a plausibly useful error message.

Consider whether there are similar low-hanging improvements nearby?